cli/src/alluxio.org/cli/processes/job_masters.go

```diff
@@ -0,0 +1,149 @@
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package processes
+
+import (
+	"path"
+
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+
+	"alluxio.org/cli/env"
+	"alluxio.org/log"
+)
+
+var JobMasters = &JobMastersProcess{
+	BaseProcess: &env.BaseProcess{
+		Name: "job_masters",
+	},
+}
+
+type JobMastersProcess struct {
+	*env.BaseProcess
+}
+
+func (p *JobMastersProcess) SetEnvVars(envVar *viper.Viper) {
+	return
+}
+
+func (p *JobMastersProcess) Base() *env.BaseProcess {
+	return p.BaseProcess
+}
+
+func (p *JobMastersProcess) StartCmd(cmd *cobra.Command) *cobra.Command {
+	cmd.Use = p.Name
+	return cmd
+}
+
+func (p *JobMastersProcess) StopCmd(cmd *cobra.Command) *cobra.Command {
+	cmd.Use = p.Name
+	return cmd
+}
+
+func (p *JobMastersProcess) Start(cmd *env.StartProcessCommand) error {
+	// get list of all masters, stored at mastersList
+	masters, err := getMasters()
+	if err != nil {
+		log.Logger.Fatalf("Cannot get masters, error: %s", err)
+	}
+
+	// get public key for passwordless ssh
+	key, err := getPrivateKey()
+	if err != nil {
+		log.Logger.Fatalf("Cannot get private key, error: %s", err)
+	}
+
+	// generate command
+	cliPath := path.Join(env.Env.EnvVar.GetString(env.ConfAlluxioHome.EnvVar), "bin", "alluxio")
+	arguments := "process start job_master"
+	if cmd.AsyncStart {
+		arguments = arguments + " -a"
+	}
+	if cmd.SkipKillOnStart {
+		arguments = arguments + " -N"
+	}
+	command := cliPath + " " + arguments
+
+	// for each master, create a client and run
+	// TODO: now start master one by one, need to do them in parallel
+	var errors []error
+	for _, master := range masters {
+		conn, err := dialConnection(master, key)
+		if err != nil {
+			errors = append(errors, err)
+			continue
+		}
+		err = runCommand(master, conn, command)
+		if err != nil {
+			errors = append(errors, err)
+		}
+		err = closeConnection(master, conn)
+		if err != nil {
+			errors = append(errors, err)
+		}
+	}
+
+	if len(errors) == 0 {
+		log.Logger.Infof("Run command %s successful on masters: %s", command, masters)
+	} else {
+		log.Logger.Fatalf("Run command %s failed, number of failures: %v", command, len(errors))
+	}
+	return nil
+}
+
+func (p *JobMastersProcess) Stop(cmd *env.StopProcessCommand) error {
+	// get list of all masters, stored at mastersList
+	masters, err := getMasters()
+	if err != nil {
+		log.Logger.Fatalf("Cannot get masters, error: %s", err)
+	}
+
+	// get public key for passwordless ssh
+	key, err := getPrivateKey()
+	if err != nil {
+		log.Logger.Fatalf("Cannot get private key, error: %s", err)
+	}
+
+	// generate command
+	cliPath := path.Join(env.Env.EnvVar.GetString(env.ConfAlluxioHome.EnvVar), "bin", "alluxio")
+	arguments := "process stop job_master"
+	if cmd.SoftKill {
+		arguments = arguments + " -s"
+	}
+	command := cliPath + " " + arguments
+
+	// for each master, create a client and run
+	// TODO: now stop master one by one, need to do them in parallel
```

Comment:
could use a `chan error` to hold all the errors and create goroutine to run the commands in parallel. it's okay to do this in a follow-up pr.
```golang
errs := make(chan error, len(masters))
go func() {
  err := runCommand()
  errs <- err // put the error to indicate the finish of the routine, whatever the error is.
}()
for i := 0; i < len(masters); i++ {
  err := <-errs
  if err == nil {
    continue
  }
  // handle err
}
```
and to make the logging more pretty the channel could accept a struct recording the hostname and the error

Reply:
agree that running a command per host is too slow. this script could potentially be launching 10 or 100 processes, so running them one at a time will not be fast enouugh.

the second section of that code should also include a way to timeout of the set of commands take too long to run, using a switch statement to handle reading from both the err channel and timeout channel